compressors/huffman_coder.py

```diff
@@ -0,0 +1,133 @@
+from dataclasses import dataclass
+from typing import Any
+from core.data_stream import BitsDataStream, DataStream
+from core.prob_dist import ProbabilityDist
+from core.data_compressor import DataCompressor
+from core.data_transformer import (
+    BitstringToBitsTransformer,
+    CascadeTransformer,
+    BitsParserTransformer,
+    LookupTableTransformer,
+)
+
+
+@dataclass
+class HuffmanTree:
+    left_child: Any = None
+    right_child: Any = None
+    id: Any = None
+    code: str = ""
+    prob: float = None
+
+    @property
+    def is_leaf_node(self):
+        return (self.left_child is None) and (self.right_child is None)
+
+    @classmethod
+    def build_huffman_tree(cls, prob_dist: ProbabilityDist):
+        """
+        Build the huffman coding tree
+        NOTE: Not the most efficient implementation. The insertion/sorting efficiency can be improved
```

Comment:
Should we switch to https://docs.python.org/3/library/heapq.html? It's clearly more efficient and might be conceptually equally as clear as the current implementation. One slight complication is in providing the sorting key to the heap thing. The best way is to have a custom comparator for the Huffman Node type (using an operator: https://docs.python.org/3/library/operator.html).
Overall, might be fine to leave it as is for now, but should be relatively straightforward to switch to a heap based implementation at some point.

Reply:
@shubhamchandak94 , I added this code!